src/detect-bytemath.c

```diff
@@ -0,0 +1,1210 @@
+
+    if (!(data->flags & DETECT_BYTEMATH_FLAG_RELATIVE))
+        goto okay;
+
+    if (prev_pm == NULL)
+        goto okay;
+
+    if (prev_pm->type == DETECT_CONTENT) {
+        DetectContentData *cd = (DetectContentData *)prev_pm->ctx;
+        cd->flags |= DETECT_CONTENT_RELATIVE_NEXT;
+    } else if (prev_pm->type == DETECT_PCRE) {
+        DetectPcreData *pd = (DetectPcreData *)prev_pm->ctx;
+        pd->flags |= DETECT_PCRE_RELATIVE_NEXT;
+    }
+
+ okay:
+    return 0;
+
+ error:
+    if (rvalue)
+        SCFree(rvalue);
+    DetectByteMathFree(de_ctx, data);
+    return ret;
+}
+
+/**
+ * \brief Used to free instances of DetectByteMathractData.
+ *
+ * \param ptr Instance of DetectByteMathData to be freed.
+ */
+static void DetectByteMathFree(DetectEngineCtx *de_ctx, void *ptr)
+{
+    if (ptr != NULL) {
+        DetectByteMathData *bmd = ptr;
+        if (bmd->result != NULL)
+            SCFree((void *)bmd->result);
+        SCFree(bmd);
+    }
+
+    return;
+}
+
+/**
+ * \brief Lookup the SigMatch for a named byte_math variable.
+ *
+ * \param arg The name of the byte_math variable to lookup.
+ * \param s Pointer the signature to look in.
+ *
+ * \retval A pointer to the SigMatch if found, otherwise NULL.
+ */
+SigMatch *DetectByteMathRetrieveSMVar(const char *arg, const Signature *s)
+{
+    const int nlists = s->init_data->smlists_array_size;
+    for (int list = 0; list < nlists; list++) {
+        SigMatch *sm = s->init_data->smlists[list];
+        while (sm != NULL) {
+            if (sm->type == DETECT_BYTEMATH) {
+                const DetectByteMathData *bmd = (const DetectByteMathData *)sm->ctx;
+                if (strcmp(bmd->result, arg) == 0) {
+                    SCLogDebug("Retrieved SM for \"%s\"", arg);
+                    return sm;
+                }
+            }
+            sm = sm->next;
+        }
+    }
+
+    return NULL;
+}
+
+/*************************************Unittests********************************/
+#ifdef UNITTESTS
+
+#include "app-layer-parser.h"
+
+static int DetectByteMathParseTest01(void)
+{
+
+    DetectByteMathData *bmd = DetectByteMathParse(NULL, "bytes 4, offset 2, oper +,"
+                                                  "rvalue 10, result bar", NULL);
+    FAIL_IF(bmd == NULL);
+
+    FAIL_IF_NOT(bmd->nbytes == 4);
+    FAIL_IF_NOT(bmd->offset == 2);
+    FAIL_IF_NOT(bmd->oper == DETECT_BYTEMATH_OPERATOR_PLUS);
+    FAIL_IF_NOT(bmd->rvalue == 10);
+    FAIL_IF_NOT(strcmp(bmd->result, "bar") == 0);
+    FAIL_IF_NOT(bmd->endian == DETECT_BYTEMATH_ENDIAN_DEFAULT);
+    FAIL_IF_NOT(bmd->base == DETECT_BYTEMATH_BASE_DEFAULT);
+
+    DetectByteMathFree(NULL, bmd);
+
+    PASS;
+}
+
+static int DetectByteMathParseTest02(void)
+{
+    /* bytes value invalid */
+    DetectByteMathData *bmd = DetectByteMathParse(NULL, "bytes 257, offset 2, oper +, "
+                                                  "rvalue 39, result bar", NULL);
+
+    FAIL_IF(bmd != NULL);
```

Comment:
`FAIL_IF_NOT_NULL(bmd)` has the same effect, but makes it even more clear that we expect the result to NULL.